godot-core/src/builtin/color_hsv.rs

```diff
@@ -0,0 +1,247 @@
+/// // are mutated both on `Color` and `ColorHsv`.
+/// let color1 = Color::from_rgb(0.74, 0.69, 0.18);
+/// let color2: Color = ColorHsv::from(color1).into();
+///
+/// let diff_r = (color1.r - color2.r).abs();
+/// let diff_g = (color1.g - color2.g).abs();
+/// let diff_b = (color1.b - color2.b).abs();
+///
+/// assert_ne!(color1, color2);
+///
+/// assert!(diff_r < 0.001);
+/// assert!(diff_g < 0.001);
+/// assert!(diff_b < 0.001);
+/// ```
+///  
+/// ## Reference
+/// - Smith, Alvy Ray. "Color gamut transform pairs." ACM Siggraph Computer Graphics 12.3 (1978): 12-19.
+#[derive(Debug, Clone, Copy, PartialOrd, Default)]
+pub struct ColorHsv {
+    h: f32,
+    s: f32,
+    v: f32,
+    a: f32,
+}
+
+// Manual implementation because Hue minimal value (0.) is equal to maximal value (1.)
+impl PartialEq for ColorHsv {
+    fn eq(&self, other: &Self) -> bool {
+        (self.h == other.h || (self.h - other.h).abs() == 1.)
+            && self.s == other.s
+            && self.v == other.v
+            && self.a == other.a
+    }
+}
+
+impl ColorHsv {
+    /// Construct from Hue, Saturation, Value where all values are between `0.0` and `1.0`.
+    ///
+    /// If values outside of that range are provided, they will be clamped to valid range. To add `alpha` value, use  `with_alpha()` method.
+    pub fn from_hsv(h: f32, s: f32, v: f32) -> Self {
+        ColorHsv {
+            h: h.clamp(0., 1.),
+            s: s.clamp(0., 1.),
+            v: v.clamp(0., 1.),
+            a: 1.,
+        }
+    }
+
+    /// Construct from Hue, Saturation, Value where Hue is represented in deegrees (`0.0` - `360.0`).
+    ///
+    /// Saturation, Value and Alpha have to be provided in range between `0.0` an `1.0`. If values outside of valid ranges are provided,
+    /// they will be clamped to a valid range. To add `alpha` value, use `with_alpha()` method.
+    pub fn from_hsv_deg(h: f32, s: f32, v: f32) -> Self {
+        ColorHsv {
+            h: h.clamp(0., 360.) / 360.,
+            s: s.clamp(0., 1.),
+            v: v.clamp(0., 1.),
+            a: 1.,
+        }
+    }
+
+    /// Adds an alpha value.
+    pub fn with_alpha(mut self, alpha: f32) -> Self {
+        self.a = alpha.clamp(0., 1.);
+        self
+    }
+
+    pub fn h(&self) -> f32 {
+        self.h
+    }
+
+    pub fn h_deg(&self) -> f32 {
+        self.h * 360.
+    }
+
+    pub fn set_h(&mut self, h: f32) {
+        self.h = h.clamp(0., 1.);
+    }
+
+    pub fn set_h_deg(&mut self, h: f32) {
+        self.h = h.clamp(0., 360.) / 360.;
+    }
+
+    pub fn s(&self) -> f32 {
+        self.s
+    }
+
+    pub fn set_s(&mut self, s: f32) {
+        self.s = s.clamp(0., 1.);
+    }
+
+    pub fn v(&self) -> f32 {
+        self.v
+    }
+
+    pub fn set_v(&mut self, v: f32) {
+        self.v = v.clamp(0., 1.);
+    }
+
+    pub fn alpha(&self) -> f32 {
+        self.a
+    }
+
+    pub fn set_alpha(&mut self, alpha: f32) {
+        self.a = alpha.clamp(0., 1.);
+    }
```

Comment:
I assume the main idea of this encapsulation is to ensure that values always stay in bounds.

While I agree that this improves invariants, most other builtin types do not have an API that follows such a design (and it's typically not enforcable because Godot can modify those). An alternative approach would also be to have assertions only in places that _use_ those invariants (e.g. conversions).

Some more input on this would be welcome 🙂

Reply:
Yeah - this is to not allow modifying values beyond the valid range. From what I've understood, the `Color` allows for values beyond that range for a reasons which wouldn't be reasonable with HSV representation. As described in `Color` doc:

> Channel values are typically in the range of 0 to 1, but this is not a requirement, and values outside this range are explicitly allowed for e.g. High Dynamic Range (HDR).